### ISSUE TITLE
When using a pre-trained model there shouldn't be dependencies on dataset steps

### ISSUE THREAD

**Issue by eu9ene** [mozilla/firefox-translations-training#395]
For example I want to experiment with teacher training and use a pre-trained backward model (very common use case). 

If I change the training data or cleaning procedure for the training corpus Taskcluster will rerun decoding with the pre-trained backward model because `merge-corpus`, `merge-devset` and `train-vocab` are upstream dependencies. This slows down experimentation a lot and wastes resources.

We should remove those dependencies from the graph if a pre-trained model is being used.

@gabrielBusta

**Comment by bhearsum** [comment]
This could also happen when re-using other pretrained teachers. The general problem seems to be that we still have some dependencies on upstream tasks that we don't use artifacts from. (Perhaps we can do some sanity checking here that we don't have dependencies on something we don't have a `fetch` for?)

**Comment by bhearsum** [comment]
@eu9ene - do you have an example of this handy? (I want to make sure I can repro locally to verify a fix...)

**Comment by bhearsum** [comment]
And could you also clarify exactly which tasks run that shouldn't? I think you're talking about `collect-mono-trg`, `translate-mono-trg`, and `split-mono-trg` when you say "rerun decoding" but I want to be certain...

**Comment by eu9ene** [comment]
I'm running this experiment with a pre-trained backward model now https://firefox-ci-tc.services.mozilla.com/tasks/Wz8KJcqPTAOFhV-YaRGJ0g. You could just add a pre-trained model to the CI config and check that it doesn't rerun translate steps by the backward model if you change something in the cleaning scripts.

I'm mostly talking about tranlsate-mono-trg because it's quite heavy and waste of resources. Currently f we modify anything upstream from merge-corpus it will rerun the fake train-backward and translate-trg and related steps.

**Comment by bhearsum** [comment]
I think I understand what's going on here. The unnecessary dependencies when we use the `use` training continuation mode is one part of it (it causes us to pull in `merge-corpus` and `merge-devset` for `train-backwards`, for example) but the bigger problem is that because we still run a `train-backwards` step, we need to satisfy all dependencies downstream of it. Most notably in this, `translate-mono-trg` and `collect-mono-trg` exist between `train-backwards` and `train-teacher`. Unless we find these in the cache, they will run. And if _their_upstream change (such as cleaning tasks), they will be forced to run again.

I'm not sure there's really a perfect fix for this. In the specific case we're looking at we happen to have cached tasks that we figure should get used -- but that's not always going to be the case. (Eg: imagine a scenario where we re-use a model a couple of years from now - there will be no cached `translate-mono-trg` tasks that we could use.) We also have no way to actually locate such tasks when using a pretrained model, as those typically come from a GCP bucket rather than a task.

One thing we _could_ possibly do is build on the work in #377 to allow explicitly replacing certain tasks with previous versions provided in the training config. This mechanism would bypass caching - so changes to things like cleaning scripts wouldn't invalidate the re-use.

**Comment by bhearsum** [comment]
The other thing I want to point out is that I think taskgraph is doing the right thing at the moment -- it notices that the cleaning scripts have changed, and wants to re-do a all tasks downstream of them because of this. What it sounds like we're looking for here is a way to bypass that (which is why it seems similar to #377 to me). Please do correct me if that's wrong though!

**Comment by eu9ene** [comment]
There are two cleaning branches: `clean-corpus` and `clean-mono`.  The first one leads to `merge-corpus` which is used for training. If there's no training of backward model, there shouldn't be dependencies on `merge-corpus` and other steps. The `clean-mono` leads to a bunch of data augmentation tasks including `translate-mono-trg`. If the monolingual data has changed it's expected that decoding by backward model will happen again. In practice we rarely change it and will mostly experiment with `clean-corpus` and related tasks to retrain teacher many times and see the difference. Anyway regardless of the scenario if the training doesn't happen, there shouldn't be any dependencies on training data. 

Basically what we want here is not bypassing the caching mechanism which works correctly as you pointed out but modifying the graph when using pre-trained models. Let's discuss live.

**Comment by bhearsum** [comment]
We chatted about this on Zoom earlier. Most of the confusing I had was around testing against existing graphs that I expected were experiencing this already - but @eu9ene cleared up that that wasn't the case. I've now got a better idea of how to verify this.

We also realized while on the call that we need to adjust `cache` parameters for the `use` mode as well -- the only thing that should impact the cache digest is the `pretrained` parameters. No other parameters or scripts should cause the cache to be invalidated. (This will fix the case where, eg: the cleaning script is changed to adjust corpus data, but at the moment will cause the entire mono dataset chain to rerun as well.)
